lib/extensions/ar_miq_set.rb

```diff
@@ -96,8 +113,42 @@ def model_table_name
     end
   end
 
-  def add_member(member)
-    raise "object of type #{member.class} may not be a member of a set of type #{self.class}" unless member.kind_of?(self.class.model_class)
-    with_relationship_type("membership") { add_child(member) }
+  def members=(*members)
+    transaction do
+      remove_all_members
+      add_members(*members)
+    end
+  end
+  alias replace_children members=
+
+  def remove_member(member)
+    miq_set_memberships.where(:member => member).delete_all
+  end
+  alias remove_child remove_member
+
+  def remove_all_members
+    miq_set_memberships.delete_all
+  end
+  alias remove_all_children remove_all_members
+
+  def add_members(*members)
+    added = []
+
+    transaction do
+      existing = miq_set_memberships.index_by { |ms| [ms.member_type, ms.member_id] }
+
+      members.flatten.each do |member|
```

Comment:
In a **followup**, does it make sense to just drop `added`?

The return value from this function is `members.flatten` which is the return from `members.flatten.each`
So if we delete `added` from the 3 places it should still work.
I don't like depending upon the return value of `each`, but I like even less building another array.

(the return value is only used in tests/factories from what I can see anyway)

Reply:
> In a followup, does it make sense to just drop added?

Sure, but I am not in the business of adding or changing functionality WHEN it is currently being used.  I don't want to burn in that risk when it could potentially be used by automate scripts and what not that our tests have no insight into.

The exception I made was here:

https://github.com/ManageIQ/manageiq/pull/21241#discussion_r640916490

Because it was not being used any where else, so it made sense to drop it before we forgot about it.

However, we could literally refactor this for years, and we wouldn't get anywhere, and these kind of things are better candidates for fixes in followups, because they don't require a schema change to be merged at the same time.